src/parser/parse_command.c

```diff
@@ -0,0 +1,41 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   parse_command.c                                    :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: reasuke <reasuke@student.42tokyo.jp>       +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2024/09/17 02:19:16 by reasuke           #+#    #+#             */
+/*   Updated: 2024/09/17 02:28:48 by reasuke          ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "ast.h"
+#include "parser_internal.h"
+#include "token.h"
+
+/*
+** command  : simple_command
+**          : '(' list ')'
+**          : '(' list ')' redirect_list
+**          ;
+*/
+t_ast	*parse_command(t_token_list **cur_token)
+{
+	t_ast	*node;
+
+	if (get_token_type(*cur_token) != TOKEN_L_PARENTHESIS)
+		return (parse_simple_command(cur_token));
+	consume_token(cur_token);
+	node = parse_list(cur_token);
+	if (node == NULL)
+		return (NULL);
+	if (!expect_token(cur_token, TOKEN_R_PARENTHESIS))
+		return (handle_error(node, get_token_value(*cur_token)));
+	if (is_redirect_token(get_token_type(*cur_token)))
+	{
+		if (!parse_redirect(node, cur_token))
```

Comment:
since `parse_redirect()` is called here and other than from `parse_simple_command()`, we must check syntax error inside `parse_redirect()` instead of in front of `parse_simple_command()`.

Reply:
this test will fail.
```
TEST(parse_command, InvalidTokenRedirect1) {
  t_token_list *token_list = nullptr;

  // (ls -l ) |
  ft_lstadd_back(&token_list,
                 construct_token(TOKEN_L_PARENTHESIS, strdup("(")));
  ft_lstadd_back(&token_list, construct_token(TOKEN_WORD, strdup("ls")));
  ft_lstadd_back(&token_list, construct_token(TOKEN_WORD, strdup("-l")));
  ft_lstadd_back(&token_list,
                 construct_token(TOKEN_R_PARENTHESIS, strdup(")")));
  ft_lstadd_back(&token_list, construct_token(TOKEN_PIPE, strdup("|")));
  ft_lstadd_back(&token_list, construct_token(TOKEN_EOF, nullptr));

  t_ast *node = parse_command(&token_list);

  EXPECT_EQ(node, nullptr);

  destroy_token_list(token_list);
}
```

Reply:
@alex216 
I think this test is not appropriate for `parse_command` but suitable for `parse_list` or `parse_pipeline`. 
`parse_command` is designed to handle only simple_command, subshell, and redirects.

To improve our test coverage, we need to add new tests specifically for parse_pipeline and parse_list to cover more complex command structures.

Reply:
To improve our test coverage, we need to add new tests for parse_pipeline and parse_list to cover more complex command structures.

Reply:
I was wrong. you are right, the test should be for `parse_pipeline()` and it works perfectly already!